clippy_lints/src/lifetimes.rs

```diff
@@ -586,11 +591,46 @@ where
         self.generic_args_depth -= 1;
     }
 
+    fn visit_fn_decl(&mut self, fd: &'tcx FnDecl<'tcx>) -> Self::Result {
+        self.lifetime_elision_impossible = !is_candidate_for_elision(fd);
+        walk_fn_decl(self, fd);
+        self.lifetime_elision_impossible = false;
+    }
+
     fn nested_visit_map(&mut self) -> Self::Map {
         self.cx.tcx.hir()
     }
 }
 
+/// Check if `fd` supports function elision with an anonymous (or elided) lifetime,
+/// and has a lifetime somewhere in its output type.
+fn is_candidate_for_elision(fd: &FnDecl<'_>) -> bool {
+    struct V;
+
+    impl Visitor<'_> for V {
+        type Result = ControlFlow<bool>;
+
+        fn visit_lifetime(&mut self, lifetime: &Lifetime) -> Self::Result {
+            ControlFlow::Break(lifetime.is_elided() || lifetime.is_anonymous())
```

Comment:
This currently breaks on the first lifetime it finds. Meaning that `(x: Cx<'_, 'tcx>)` would say it's anonymous, right?

It feels like this should only break if it finds a user-defined lifetime :thinking: 

Could you also add a test for this?

Reply:
Yes, but:

- when checking the output type, we don't care about the lifetime being anonymous (we just want to know if we have at least one lifetime)
- we check the input types only when the function supports lifetime elision, which is not the case here because we have at least two input lifetimes (`'_` and `'tcx`)

Does that make sense? I've added a test, let me know if this is what you had in mind.